File: 백준/Gold/1806. 부분합/부분합.java
import java.io.*;
import java.util.*;

public class Main{

    static int N, S;
    static int[] nums;
    static void input() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        S = Integer.parseInt(st.nextToken());

        // 수열 입력 받기
        st = new StringTokenizer(br.readLine());
        nums = new int[N + 1];
        for (int i = 1; i <= N ; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        br.close();
    }

    static void pro() {
        int r = 0, sum = 0, ans = N + 1;
        for (int l = 1; l <= N; l++) {
            // L - 1을 구간에서 제외하기
            sum -= nums[l - 1];

            // R을 옮길 수 잇을 때 까지 옮기기
            while (r + 1 <= N && sum < S) {
                sum += nums[r + 1];
                r++;
            }

            // [L...R]의 합, 즉 sum이 조건을 만족하면 정답 갱신하기
            if (sum >= S) {
                ans = Math.min(ans, r - l + 1);
            }
        }

        // ans 값을 보고 불가능 판단하기
        if (ans == N + 1) {
            ans = 0;
        }
        System.out.println(ans);
    }
    public static void main(String[] args) throws Exception{
        input();
        pro();
    }
}

File: 백준/Gold/1806. 부분합/부분합.cc
#include <iostream>
#include <vector>

using namespace std;

// 변수
int N, S;
vector<int> nums;

void input() {
    cin >> N >> S;

    int temp;
    nums.push_back(0);
    for (int i = 0; i < N; i++) {
        cin >> temp;
        nums.push_back(temp);
    }

}

void pro() {
    int R = 1, ans = N + 1, sum = nums[1];

    for (int L = 1; L <= N; L++) {
        // L - 1을 빼기
        sum -= nums[L - 1];

        // S이상이 될때까지 더하기(R이동)
        while (sum < S && R + 1 <= N) {
            sum += nums[++R];
        }

        // 길이비교
        if (sum >= S) {
            ans = min(ans, R - L + 1);
        }
    }

    // 0인지 체크
    if (ans == N + 1) {
        ans = 0;
    }

    // 출력
    cout << ans;
}

int main() {
    // 최적화
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    // 입력
    input();

    // 실행
    pro();
}